File: js/49-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
  const result = new Object();
  const output_array = [];

  strs.forEach(str => {
    const sort_str = str.split("").sort().join("");
    result[sort_str] ||= []
    result[sort_str].push(str)
  });
  Object.keys(result).map(key => output_array.push(result[key]));
  return output_array;
};

console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));
console.log(groupAnagrams([""]));
console.log(groupAnagrams(["a"]));

File: ruby/49-group-anagrams.rb
# @param {String[]} strs
# @return {String[][]}
def group_anagrams(strs)
  grouped = {}

  strs.each do |str|
    key = str.chars.sort.join
    if grouped.key? key
      grouped[key] << str
    else
      grouped[key] = [str]
    end
    grouped
  end
  grouped.values  
end

p group_anagrams(["eat","tea","tan","ate","nat","bat"])
p group_anagrams([""])
p group_anagrams(["a"])
